### PR TITLE
snap/naming: introduce SnapRef, Snap, and SnapSet

### PR DESCRIPTION
**Pull request by pedronis**
This introduces the concept of a SnapRef(erence)  with a new interface. The main goal here is to be able to write code that can work both in cases where a snap id is known for some snaps or the name only.

This will now be the case for code that deals with model assertions, so far they carried only names but for Core 20 they will carry snap ids for the grade stable at least.

This is introduces also SnapSet to be able to work with a set of these.

### CODE REVIEW

File: snap/naming/snapref.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package naming
+
+// A SnapRef references a snap by name and/or id.
```
Comment (zyga) [inline snap/naming/snapref.go]: I think it is worth to |spell out that the interface guarantees that either name or id is present.

File: snap/naming/snapref.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	return s.byNameOnly[name]
+}
+
+// Contains returns whether the set has a matching reference already.
+func (s *SnapSet) Contains(ref SnapRef) bool {
+	return s.Lookup(ref) != nil
+}
+
+// Add adds one reference to the set. Already added ids or names will be ignored. The assumption is that a SnapSet is populated with distinct snaps.
```
Comment (zyga) [inline snap/naming/snapref.go]:
```suggestion
// Add adds one reference to the set.
// Already added ids or names will be ignored. The assumption is that a SnapSet is populated with distinct snaps.
```

File: snap/naming/snapref.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	id1 := snapRef1.ID()
+	id2 := snapRef2.ID()
+	if id1 != "" && id2 != "" {
+		return id1 == id2
+	}
+	return snapRef1.SnapName() == snapRef2.SnapName()
+}
+
+// SnapSet can hold a set of references to snaps.
+type SnapSet struct {
```
Comment (zyga) [inline snap/naming/snapref.go]:
```suggestion
// No provisions are made for correctness in concurrent acess.
type SnapSet struct {
```

File: snap/naming/snapref.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package naming
+
+// A SnapRef references a snap by name and/or id.
+type SnapRef interface {
+	SnapName() string
```
Comment (zyga) [inline snap/naming/snapref.go]: Is this always snap name and not snap instance name?

File: snap/naming/snapref.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+// SnapSet can hold a set of references to snaps.
+type SnapSet struct {
+	byID         map[string]SnapRef
+	byNameWithID map[string]SnapRef
+	byNameOnly   map[string]SnapRef
+}
+
+// NewSnapSet builds a snap set with the given references.
+func NewSnapSet(refs []SnapRef) *SnapSet {
```
Comment (chipaca) [inline snap/naming/snapref.go]: I'm curious about how this is going to be used: it seems memory-inefficient for something long-lived, and speed-inefficient for something short-lived. So...
Granted it's a first pass to get the implementation right etc etc, but was still wondering.

File: snap/naming/snapref.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+type snapRef struct {
+	name string
+	id   string
+}
+
+// NewSnapRef returns a reference to the snap with given name and id.
+func NewSnapRef(name, id string) SnapRef {
+	return &snapRef{name: name, id: id}
```
[the remark this answers is not in the document]
Comment (pedronis) [reply snap/naming/snapref.go]: error is overkill. SnapSet will not add a ("", "") ref. Likely if we have that situation some other bit will fail, this in the end is just a carrier of two strings.

File: snap/naming/snapref.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+type SnapSet struct {
+	byID         map[string]SnapRef
+	byNameWithID map[string]SnapRef
+	byNameOnly   map[string]SnapRef
+}
+
+// NewSnapSet builds a snap set with the given references.
+func NewSnapSet(refs []SnapRef) *SnapSet {
+	s := &SnapSet{
+		byID:         make(map[string]SnapRef),
```
Comment (bboozzoo) [inline snap/naming/snapref.go]: Wonder if it'd be easier to have:
```
type SnapSet struct {
   byID      map[string]Snap  // key is snap ID, val is snap name
   byName    map[Snap]string  // key is snap name, val is snap ID or empty if ID is not known
}
```

Comment (pedronis) [reply snap/naming/snapref.go]: I do need to keep the actual SnapRefs for Lookup  vs Contains, but yes there should be a way to have only two maps and not 3.

**Comment by pedronis** [comment]
Merging and I will see if I can simplify SnapSet a bit in a follow up

**Pull request merged**